log4j-api/src/main/java/org/apache/logging/log4j/ScopedContext.java

```diff
@@ -0,0 +1,558 @@
+            } else {
+                new Instance(map).run(op);
+                return CompletableFuture.completedFuture(0);
+            }
+        } else {
+            Map<String, Renderable> map = new HashMap<>();
+            if (current().isPresent()) {
+                map.putAll(current().get().contextMap);
+            }
+            map.remove(key);
+            if (executorService != null) {
+                return executorService.submit(new Runner(
+                        new Instance(map), ThreadContext.getContext(), ThreadContext.getImmutableStack(), op));
+            } else {
+                new Instance(map).run(op);
+                return CompletableFuture.completedFuture(0);
+            }
+        }
+    }
+
+    /**
+     * Creates a ScopedContext with a Map of keys and values and calls a method.
+     * @param map the Map.
+     * @param op the Runnable to call.
+     */
+    public static void runWhere(Map<String, ?> map, Runnable op) {
+        if (map != null && !map.isEmpty()) {
+            Map<String, Renderable> renderableMap = new HashMap<>();
+            if (current().isPresent()) {
+                renderableMap.putAll(current().get().contextMap);
+            }
+            map.forEach((key, value) -> {
+                renderableMap.put(key, value instanceof Renderable ? (Renderable) value : new ObjectRenderable(value));
+            });
+            new Instance(renderableMap).run(op);
+        } else {
+            op.run();
+        }
+    }
+
+    /**
+     * Creates a ScopedContext with a single key/value pair and calls a method.
+     * @param key the key.
+     * @param obj the value associated with the key.
+     * @param op the Runnable to call.
+     */
+    public static <R> R callWhere(String key, Object obj, Callable<R> op) throws Exception {
+        if (obj != null) {
+            Renderable renderable = obj instanceof Renderable ? (Renderable) obj : new ObjectRenderable(obj);
+            Map<String, Renderable> map = new HashMap<>();
+            if (current().isPresent()) {
+                map.putAll(current().get().contextMap);
+            }
+            map.put(key, renderable);
+            return new Instance(map).call(op);
+        } else {
+            Map<String, Renderable> map = new HashMap<>();
+            if (current().isPresent()) {
+                map.putAll(current().get().contextMap);
+            }
+            map.remove(key);
+            return new Instance(map).call(op);
+        }
+    }
+
+    /**
+     * Creates a ScopedContext with a single key/value pair and calls a method on a separate Thread.
+     * @param key the key.
+     * @param obj the value associated with the key.
+     * @param executorService the ExecutorService to dispatch the work.
+     * @param op the Callable to call.
+     */
+    public static <R> Future<R> callWhere(String key, Object obj, ExecutorService executorService, Callable<R> op)
```

Comment:
We've got Java 8 at minimum at our fingertips. Let's make sure to support `CompletableFuture` as well. If possible, it'd be great to support `Flow.Publisher` from Java 9, too, though context propagation with reactive streams tends to use a non-standard API such as Reactor's context.

Reply:
@jvz Isn't CompletableFuture a Future? How is it not supported now?

Reply:
It implements Future, yes, but it's also a CompletionStage which is the part of the API I'm more interested in.